runtime/interpreter/value.go

```diff
@@ -2695,247 +2692,1468 @@ type NumberValue interface {
-			func(invocation Invocation) Value {
-				other, ok := invocation.Arguments[0].(NumberValue)
-				if !ok {
-					panic(errors.NewUnreachableError())
-				}
-				return v.SaturatingPlus(
-					invocation.Interpreter,
-					other,
-					locationRange,
-				)
-			},
-			&sema.FunctionType{
-				ReturnTypeAnnotation: sema.NewTypeAnnotation(
-					typ,
-				),
-			},
-		)
-
-	case sema.NumericTypeSaturatingSubtractFunctionName:
-		return NewHostFunctionValue(
-			interpreter,
-			func(invocation Invocation) Value {
-				other, ok := invocation.Arguments[0].(NumberValue)
-				if !ok {
-					panic(errors.NewUnreachableError())
-				}
-				return v.SaturatingMinus(
-					invocation.Interpreter,
-					other,
-					locationRange,
-				)
-			},
-			&sema.FunctionType{
-				ReturnTypeAnnotation: sema.NewTypeAnnotation(
-					typ,
-				),
-			},
-		)
-
-	case sema.NumericTypeSaturatingMultiplyFunctionName:
-		return NewHostFunctionValue(
-			interpreter,
-			func(invocation Invocation) Value {
-				other, ok := invocation.Arguments[0].(NumberValue)
-				if !ok {
-					panic(errors.NewUnreachableError())
-				}
-				return v.SaturatingMul(
-					invocation.Interpreter,
-					other,
-					locationRange,
-				)
-			},
-			&sema.FunctionType{
-				ReturnTypeAnnotation: sema.NewTypeAnnotation(
-					typ,
-				),
-			},
-		)
+type Unsigned interface {
+	constraints.Unsigned | ~*big.Int
+}
 
-	case sema.NumericTypeSaturatingDivideFunctionName:
-		return NewHostFunctionValue(
-			interpreter,
-			func(invocation Invocation) Value {
-				other, ok := invocation.Arguments[0].(NumberValue)
-				if !ok {
-					panic(errors.NewUnreachableError())
-				}
-				return v.SaturatingDiv(
-					invocation.Interpreter,
-					other,
-					locationRange,
-				)
-			},
-			&sema.FunctionType{
-				ReturnTypeAnnotation: sema.NewTypeAnnotation(
-					typ,
-				),
-			},
-		)
-	}
+type Signed interface {
+	constraints.Signed | ~*big.Int
+}
 
-	return nil
+type NumericalWithoutBigInt interface {
+	constraints.Unsigned | constraints.Signed
 }
 
-type IntegerValue interface {
-	NumberValue
-	BitwiseOr(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue
-	BitwiseXor(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue
-	BitwiseAnd(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue
-	BitwiseLeftShift(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue
-	BitwiseRightShift(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue
+type Numerical interface {
+	Unsigned | Signed
 }
 
-// BigNumberValue is a number value with an integer value outside the range of int64
-type BigNumberValue interface {
+type BoundedValue[T Numerical] interface {
 	NumberValue
-	ByteLength() int
-	ToBigInt(memoryGauge common.MemoryGauge) *big.Int
+	MaxValue() T
+	Underlying() T
+	Constructor(common.MemoryGauge, func() T) NumberValue
```

Comment:
`Constructor` sounds like this function returns a constructor (function). Maybe rename to a verb, e.g. `Construct`.

This and the other two functions `MinValue` and `MaxValue` are really don't really need the value as a receiver. Maybe there is a more efficient way to dispatch just on the type instead of the value, i.e. have something like a static function, but I can't think of a way to do that in Go 🤔